### ISSUE TITLE
Deleting ~/.stack and leaving .stack-work leads to corrupted package database

### ISSUE THREAD

**Issue by wolftune** [commercialhaskell/stack#772]
running `stack build` for [Snowdrift](https://github.com/snowdriftcoop/snowdrift)

I get this weird error:

```
--  While building package libravatar-0.1.0.1 using:
      /usr/local/bin/runhaskell -package=Cabal-1.18.1.5 -clear-package-db -global-package-db -package-db=/home/aaron/.stack/snapshots/x86_64-linux/lts-2.21/7.8.4/pkgdb/ /tmp/stack2717/Setup.hs --builddir=.stack-work/dist/x86_64-linux/Cabal-1.18.1.5/ build --ghc-options -hpcdir .stack-work/dist/x86_64-linux/Cabal-1.18.1.5/hpc/.hpc/ -ddump-hi -ddump-to-file
    Process exited with code: ExitFailure 1
    Logs have been written to: /home/aaron/sites-programs/snowdrift/.stack-work/logs/libravatar-0.1.0.1.log

    Configuring libravatar-0.1.0.1...
    Building libravatar-0.1.0.1...
    Preprocessing library libravatar-0.1.0.1...

    src/Network/Libravatar.hs:30:8:
        Could not find module ‘Network.DNS.Lookup’
        There are files missing in the ‘dns-2.0.0’ package,
        try running 'ghc-pkg check'.
        Use -v to see a list of the files searched for.

    src/Network/Libravatar.hs:31:8:
        Could not find module ‘Network.DNS.Resolver’
        There are files missing in the ‘dns-2.0.0’ package,
        try running 'ghc-pkg check'.
        Use -v to see a list of the files searched for.

    src/Network/Libravatar.hs:32:8:
        Could not find module ‘Network.DNS.Types’
        There are files missing in the ‘dns-2.0.0’ package,
        try running 'ghc-pkg check'.
        Use -v to see a list of the files searched for.
```

I had a similar error earlier (see http://paste.rel4tion.org/45 ) and tried changing stack.yaml to lts-2.22 and successfully built. I then tried removing all of .stack and starting fresh back with lts-2.21 and ran into this new error. I don't know exactly what's going on. Is this actually an issue with the LTS version and we should update to lts-2.22? How does one parse this error / troubleshoot / make sense of what is going on here?

**Comment by snoyberg** [comment]
I'm unable to reproduce. If there's not a smaller test case that demonstrates this, it's going to be very difficult to track down.

**Comment by wolftune** [comment]
@snoyberg thanks for the reply. Is there somewhere a documentation about what things need to be purged in order to have a completely fresh start if something goes wrong? Is just removing ~/.stack enough? What about other stuff in the project directory? If there's no clear documentation like that, it would be useful to make.

Recently, I did one interesting thing: I was asked to test a Stack update, so I did `stack upgrade --git` and added PATH for ~/.local/bin — but then after my test and the update getting into the release, I got an ugrade via apt-get. So then I realized I needed to remove the .local/bin version and/or the PATH in order to use the new system version of Stack. I'm not sure if something about that should be documented or it's just something people who run stack upgrade manually need to manage understanding…

So, the Stack upgrade happened, and we also updated to lts-2.21 recently. It was in the midst of this that `stack build` had its errors. Now, since I successfully built with lts-2.22, I could go back to that potentially. But I did want to at least test that I can't get it working with lts-2.21.

So, in short: I can try again, and just want to know if there's other things I should reset besides removing ~/.stack/  — if I do a full reset of things and still doesn't work, we'll move to lts-2.22 and hope that does it. I don't know how else to isolate the issue given that the errors don't seem meaningful enough to you even.

Thanks for all your work

**Comment by snoyberg** [comment]
Did you try deleting .stack-work as well?

**Comment by wolftune** [comment]
No, I forgot about .stack-work as that's only within the project directory, that makes sense, I'll try that

**Comment by wolftune** [comment]
That seems to have done it! That time, I removed .stack-work but left .stack

I think I've *also* isolated the bug in this case:

* when LTS version changes in a project, `stack build` knows to update .stack but not to update .stack-work or at least not the items within .stack-work that are extra-deps — yet those need updating because they may rely on items within LTS, otherwise this opaque bug occurs.

In other words, this bug happens whenever .stack-work has extra-deps that were built with a different LTS than current…

Certainly this should be fixed and could be tested with a very light test repo. It's possible that the bug even happens without any extra-deps as long as the project itself has any libraries or anything that go into .stack-work … surely you know more about that than I.

Cheers

**Comment by snoyberg** [comment]
I think this may be simpler than the new title: the point is that the package database in .stack-work (the local database) depends on the database in ~/.stack (the snapshot database), and deleting the latter out from under the former leads to a corrupted database. I'm on the fence right now between calling this a bug or expected behavior with a documentation fix (if you delete ~/.stack, you need to delete .stack-work too). If there's an easy way for stack to detect this situation, we should, otherwise I think documentation is sufficient.

@borsboom any thoughts?

**Comment by wolftune** [comment]
Ah, that makes sense, although I don't recall from my process if the issue actually started with removing ~/.stack

I think I *might* have hit the bug right away the first time I tried to build after the lts version change. I don't know why I would have removed ~/.stack otherwise. So I think I just updated to new master, the new master had new lts version in stack.yaml and then I hit this bug without my doing anything other than `stack build`. So, that could be tested with a smaller repo to see if this happens.

If so, it's definitely a bug if people have to manually remove .stack-work every time lts version is changed.

**Comment by snoyberg** [comment]
I've tested that case many times, in fact I did so just today. If you can
provide a step by step repro showing that add the problem, I can
investigate that.

**Comment by wolftune** [comment]
@snoyberg okay, let's just assume that for some reason I did delete .stack and that caused the error.

I can do the documentation potentially. Which wiki page would be best for mentioning how to reset stack (where I can say that one must be sure to remove both .stack and .stack-work ) ?

**Comment by snoyberg** [comment]
How about the FAQ

**Comment by wolftune** [comment]
Did a draft for now, not sure if it's complete, but it's good enough for now. Closing this

**Comment by snoyberg** [comment]
Thanks!
